src/dawn/Optimizer/MultiStage.cpp

```diff
@@ -105,21 +108,116 @@ std::shared_ptr<DependencyGraphAccesses> MultiStage::getDependencyGraphOfAxis()
 }
 
 Cache& MultiStage::setCache(Cache::CacheTypeKind type, Cache::CacheIOPolicy policy, int AccessID,
-                            const Interval& interval) {
-  return caches_
-      .emplace(AccessID, Cache(type, policy, AccessID, boost::optional<Interval>(interval)))
+                            const Interval& interval, boost::optional<Cache::window> w) {
+  return caches_.emplace(AccessID,
+                         Cache(type, policy, AccessID, boost::optional<Interval>(interval), w))
       .first->second;
 }
 
 Cache& MultiStage::setCache(Cache::CacheTypeKind type, Cache::CacheIOPolicy policy, int AccessID) {
-  return caches_.emplace(AccessID, Cache(type, policy, AccessID, boost::optional<Interval>()))
+  return caches_.emplace(AccessID, Cache(type, policy, AccessID, boost::optional<Interval>(),
+                                         boost::optional<Cache::window>()))
       .first->second;
 }
 
-boost::optional<Interval> MultiStage::computeEnclosingAccessInterval(const int accessID) const {
+std::vector<DoMethod> MultiStage::computeOrderedDoMethods() const {
+  auto intervals_set = getIntervals();
+  std::vector<Interval> intervals_v;
+  std::copy(intervals_set.begin(), intervals_set.end(), std::back_inserter(intervals_v));
+
+  // compute the partition of the intervals
+  auto partitionIntervals = Interval::computePartition(intervals_v);
+  if((getLoopOrder() == LoopOrderKind::LK_Backward))
+    std::reverse(partitionIntervals.begin(), partitionIntervals.end());
+
+  std::vector<DoMethod> orderedDoMethods;
+
+  for(auto interval : partitionIntervals) {
+    for(const auto& stagePtr : getStages()) {
+      for(const auto& doMethod : stagePtr->getDoMethods()) {
+
+        if(doMethod->getInterval().overlaps(interval)) {
+          DoMethod partitionedDoMethod(*doMethod);
+
+          partitionedDoMethod.setInterval(interval);
+          orderedDoMethods.push_back(partitionedDoMethod);
+          // there should not be two do methods in the same stage with overlapping intervals
+          continue;
+        }
+      }
+    }
+  }
+
+  return orderedDoMethods;
+}
+
+MultiInterval MultiStage::computeReadAccessInterval(int accessID) const {
+
+  std::vector<DoMethod> orderedDoMethods = computeOrderedDoMethods();
+
+  MultiInterval writeInterval;
+  MultiInterval writeIntervalPre;
+  MultiInterval readInterval;
+
+  for(const auto& doMethod : orderedDoMethods) {
+    for(const auto& statementAccesssPair : doMethod.getStatementAccessesPairs()) {
+      const Accesses& accesses = *statementAccesssPair->getAccesses();
+      if(accesses.hasWriteAccess(accessID)) {
+        writeIntervalPre.insert(doMethod.getInterval());
```

Comment:
Just to confirm: The  field-offsets are already handled? (Is the doMethod Interval extended to how large the actual acesses are?)

Reply:
uhmm didnt understand, we will have to discuss it